File: app/src/main/java/id/rofiqoff/examplefilterdataactivityfragment/java/DataFragment.java
package id.rofiqoff.examplefilterdataactivityfragment.java;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import id.rofiqoff.examplefilterdataactivityfragment.Data;
import id.rofiqoff.examplefilterdataactivityfragment.DataViewHolder;
import id.rofiqoff.examplefilterdataactivityfragment.R;
import id.rofiqoff.examplefilterdataactivityfragment.Adapter;

import static id.rofiqoff.examplefilterdataactivityfragment.DataKt.getDatas;

public class DataFragment extends Fragment {

    private RecyclerView listData;

    private Adapter<Data, DataViewHolder> adapter;

    private View itemView;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        itemView = inflater.inflate(R.layout.fragment_data, container, false);
        return itemView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView();
    }

    private void initView() {
        listData = itemView.findViewById(R.id.list_data);

        initAdapter();

        listData.setAdapter(adapter);

        ((MainActivity) getActivity()).setDataSearch(new MainActivity.SearchListener() {
            @Override
            public void onSearch(String dataToString) {
                search(dataToString);
            }
        });
    }

    private void initAdapter() {
        adapter = new Adapter<Data, DataViewHolder>(R.layout.row_list_data, DataViewHolder.class, getDatas()) {
            @Override
            protected void bindView(@org.jetbrains.annotations.Nullable DataViewHolder holder, Data model, int position) {
                holder.onBind(model);
            }
        };
    }

    private void search(String dataToString) {
        List<Data> dataFilter = new ArrayList<Data>();

        for (int i = 0; i < dataFilter.size() - 1; i++) {
            String textData = getDatas().get(i).getTextData();

            if (dataToString.contains(textData.toLowerCase())) {
                dataFilter.add(new Data(textData));
            }
        }

        if (dataToString == "") {
            adapter.updateData(getDatas());
        } else {
            adapter.updateData(dataFilter);
        }

    }
}

File: app/src/main/java/id/rofiqoff/examplefilterdataactivityfragment/DataFragment.kt
package id.rofiqoff.examplefilterdataactivityfragment


import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import kotlinx.android.synthetic.main.fragment_data.*

/**
 * A simple [Fragment] subclass.
 */
class DataFragment : Fragment() {

    private var adapter: Adapter<Data, DataViewHolder>? = null

    val datas: ArrayList<Data>
        get() = arrayListOf<Data>().apply {
            add(Data("Real Madrid"))
            add(Data("Barcelona"))
            add(Data("Manchester United"))
            add(Data("Manchester City"))
            add(Data("Liverpool"))
            add(Data("Juventus"))
            add(Data("Chelsea"))
            add(Data("Atletico Madrid"))
            add(Data("Paris Saint German"))
            add(Data("Ac Milan"))
            add(Data("Inter Milan"))
            add(Data("AS Roma"))
        }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_data, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
    }

    private fun initView() {
        initAdapter()

        list_data?.adapter = adapter

        adapter?.updateData(datas)

        (activity as MainActivity).setSearchDataListener(object :
            MainActivity.SearchDataListener {
            override fun onSearch(dataToSearch: String) {
                search(dataToSearch)
            }
        })
    }

    private fun initAdapter() {
        adapter = object : Adapter<Data, DataViewHolder>(
            R.layout.row_list_data, DataViewHolder::class.java, arrayListOf()
        ) {
            override fun bindView(holder: DataViewHolder?, model: Data, position: Int) {
                holder?.onBind(model)
            }
        }
    }

    fun search(dataToSearch: String) {
        val dataFilter = datas.filter {
            it.textData.contains(dataToSearch, true)
        }

        adapter?.apply {
            if (dataToSearch == "") {
                updateData(datas)
            } else {
                updateData(dataFilter as MutableList<Data>)
            }
        }
    }

}
